Sources/SwiftProtobuf/AsyncMessageSequence.swift

```diff
@@ -0,0 +1,202 @@
+  ///   - extensions: An `ExtensionMap` used to look up and decode any extensions in
+  ///    messages encoded by this sequence, or in messages nested within these messages.
+  ///   - partial: If `false` (the default),  after decoding a message, `Message.isInitialized`
+  ///     will be checked to ensure all fields are present. If any are missing,
+  ///     `BinaryDecodingError.missingRequiredFields` will be thrown.
+  ///   - options: The BinaryDecodingOptions to use.
+  /// - Returns: An asynchronous sequence of messages read from the `AsyncSequence` of bytes.
+  /// - Throws: `BinaryDecodingError` if decoding fails, throws
+  ///           `BinaryDelimited.Error` for some reading errors,
+  ///           `BinaryDecodingError.truncated` if the stream ends before fully decoding a
+  ///           message or a delimiter,
+  ///           `BinaryDecodingError.malformedProtobuf`if a delimiter could not be read and
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  public init(
+    base: Base,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) {
+    self.base = base
+    self.extensions = extensions
+    self.partial = partial
+    self.options = options
+  }
+  
+  /// An asynchronous iterator that produces the messages of this asynchronous sequence
+  public struct AsyncIterator: AsyncIteratorProtocol {
+    private var iterator: Base.AsyncIterator?
+    private let extensions: ExtensionMap?
+    private let partial: Bool
+    private let options: BinaryDecodingOptions
+    
+    init(
+      iterator: Base.AsyncIterator,
+      extensions: ExtensionMap?,
+      partial: Bool,
+      options: BinaryDecodingOptions
+    ) {
+      self.iterator = iterator
+      self.extensions = extensions
+      self.partial = partial
+      self.options = options
+    }
+    
+    /// Aysnchronously reads the next varint
+    private mutating func nextVarInt() async throws -> UInt64? {
+      var messageSize: UInt64 = 0
+      var shift: UInt64 = 0
+      
+      while let byte = try await iterator?.next() {
+        messageSize |= UInt64(byte & 0x7f) << shift
+        shift += UInt64(7)
+        if shift > 35 {
+          iterator = nil
+          throw BinaryDecodingError.malformedProtobuf
+        }
+        if (byte & 0x80 == 0) {
+          return messageSize
+        }
+      }
+      if (shift > 0) {
+        // The stream has ended inside a varint.
+        iterator = nil
+        throw BinaryDecodingError.truncated
+      }
+      return nil // End of stream reached.
+    }
+    
+    /// Asynchronously advances to the next message and returns it, or ends the
+    /// sequence if there is no next message.
+    ///
+    /// - Returns: The next message, if it exists, or `nil` to signal the end of
+    ///   the sequence.
+    public mutating func next() async throws -> M? {
```

Comment:
We want this method to be `@inlinable` which means we have to mark a bunch of other things here as `@usableFromInline`

Reply:
I went ahead and made `nextVarInt` `@inlinable`  too

Reply:
Also @FranzBusch would you be so kind as to elaborate on why we "want" this method to be inlinable? I ask because at first blush, I am seeing this code run a byte buffer through the underlying `iterator` byte-by-byte. So I wonder if there is any data to suggest that `@inlinable` is actually a performance win, in this case. With a baseline opinion of "inlinable is a complicated promise to make and should only be done for a well understood benefit."

Reply:
This has been a learning in the swift on server ecosystem. Basically any type that has a generic parameter benefits heavily from `@inlinable` on its method since the compiler is going to specialise the generic code across module boundaries. Of course, this often should be proven with benchmarking code but we have learned over the years that exactly this pattern is benefiting heavily from `@inlinable` annotations.

The good thing is that we can always remove the `@inlinable` annotation again without breaking any API.

Reply:
I would love to see some measurements to justify whether the increased code size is warranted in this particular case.  It's not obvious to me either way.

But I don't think this should block landing this particular PR.  So I'm going to mark this "resolved" for now.